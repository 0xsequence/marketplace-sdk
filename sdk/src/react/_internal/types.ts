import type { Address } from 'viem';
import { z } from 'zod';
import type { ContractType, CreateReq } from '../../types';
import type { MarketplaceKind } from './api';

export const QueryArgSchema = z
	.object({
		enabled: z.boolean().optional(),
	})
	.optional();

export type QueryArg = z.infer<typeof QueryArgSchema>;

export const CollectableIdSchema = z.string().or(z.number());

export const AddressSchema = z.string().transform((val, ctx) => {
	const regex = /^0x[a-fA-F0-9]{40}$/;

	if (!regex.test(val)) {
		ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: `Invalid Address ${val}`,
		});
	}

	return val as Address;
});

export enum StoreType {
	MARKETPLACE = 'marketplace',
	SHOP = 'shop',
}

export type CollectionType = ContractType.ERC1155 | ContractType.ERC721;

type TransactionStep = {
	exist: boolean;
	isExecuting: boolean;
	execute: () => Promise<void>;
};

export type TransactionSteps = {
	approval: TransactionStep;
	transaction: TransactionStep;
};

export enum TransactionType {
	BUY = 'BUY',
	SELL = 'SELL',
	LISTING = 'LISTING',
	OFFER = 'OFFER',
	TRANSFER = 'TRANSFER',
	CANCEL = 'CANCEL',
}

export interface BuyInput {
	orderId: string;
	collectableDecimals: number;
	marketplace: MarketplaceKind;
	quantity: string;
}

export interface SellInput {
	orderId: string;
	marketplace: MarketplaceKind;
	quantity?: string;
}

export interface ListingInput {
	contractType: ContractType;
	listing: CreateReq;
}

export interface OfferInput {
	contractType: ContractType;
	offer: CreateReq;
}

export interface CancelInput {
	orderId: string;
	marketplace: MarketplaceKind;
}
