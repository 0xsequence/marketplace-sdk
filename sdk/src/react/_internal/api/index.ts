export * from './get-query-client';
export * from './marketplace.gen';
export * from './marketplace-api';
export * from './query-keys';
export * from './services';
