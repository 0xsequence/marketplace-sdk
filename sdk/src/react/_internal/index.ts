export * from './api';
export * from './consts';
export * from './get-provider';
export * from './types';
export * from './wagmi';
