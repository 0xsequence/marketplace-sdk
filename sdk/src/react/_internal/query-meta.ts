export const persistentQueryMeta = {
	persistent: true,
};
