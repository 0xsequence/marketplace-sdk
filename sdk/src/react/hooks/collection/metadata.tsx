'use client';

import { useQuery } from '@tanstack/react-query';
import type { Optional } from '../../_internal';
import {
	type CollectionQueryOptions,
	collectionQueryOptions,
	type FetchCollectionParams,
} from '../../queries/collection/metadata';
import { useConfig } from '../config/useConfig';

export type UseCollectionMetadataParams = Optional<
	CollectionQueryOptions,
	'config'
>;

/**
 * Hook to fetch collection information from the metadata API
 *
 * Retrieves basic contract information including name, symbol, type,
 * and extension data for a given collection contract.
 *
 * @param params - Configuration parameters
 * @param params.chainId - The chain ID (must be number, e.g., 1 for Ethereum, 137 for Polygon)
 * @param params.collectionAddress - The collection contract address
 * @param params.query - Optional React Query configuration
 *
 * @returns Query result containing contract information
 *
 * @example
 * Basic usage:
 * ```typescript
 * const { data, isLoading } = useCollectionMetadata({
 *   chainId: 137,
 *   collectionAddress: '0x...'
 * })
 * ```
 *
 * @example
 * With custom query options:
 * ```typescript
 * const { data, isLoading } = useCollectionMetadata({
 *   chainId: 1,
 *   collectionAddress: '0x...',
 *   query: {
 *     refetchInterval: 30000,
 *     enabled: userWantsToFetch
 *   }
 * })
 * ```
 */
export function useCollectionMetadata(params: UseCollectionMetadataParams) {
	const defaultConfig = useConfig();

	const { config = defaultConfig, ...rest } = params;

	const queryOptions = collectionQueryOptions({
		config,
		...rest,
	});

	return useQuery({
		...queryOptions,
	});
}

export { collectionQueryOptions };

export type { FetchCollectionParams, CollectionQueryOptions };
