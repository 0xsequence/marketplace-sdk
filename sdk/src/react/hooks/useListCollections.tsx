'use client';

import { useQuery } from '@tanstack/react-query';
import type { Optional } from '../_internal';
import {
	type FetchListCollectionsParams,
	type ListCollectionsQueryOptions,
	listCollectionsQueryOptions,
} from '../queries/listCollections';
import { useConfig } from './useConfig';
import { useMarketplaceConfig } from './useMarketplaceConfig';

export type UseListCollectionsParams = Optional<
	ListCollectionsQueryOptions,
	'config' | 'marketplaceConfig'
>;

/**
 * Hook to fetch collections from marketplace configuration
 *
 * Retrieves all collections configured in the marketplace, with optional filtering
 * by marketplace type. Combines metadata from the metadata API with marketplace
 * configuration to provide complete collection information.
 *
 * @param params - Configuration parameters
 * @param params.marketplaceType - Optional filter by marketplace type
 * @param params.query - Optional React Query configuration
 *
 * @returns Query result containing array of collections with metadata
 *
 * @example
 * Basic usage:
 * ```typescript
 * const { data: collections, isLoading } = useListCollections();
 *
 * if (isLoading) return <div>Loading collections...</div>;
 *
 * return (
 *   <div>
 *     {collections?.map(collection => (
 *       <div key={collection.itemsAddress}>
 *         {collection.name}
 *       </div>
 *     ))}
 *   </div>
 * );
 * ```
 *
 * @example
 * Filtering by marketplace type:
 * ```typescript
 * const { data: marketCollections } = useListCollections({
 *   marketplaceType: 'market'
 * });
 * ```
 */
export function useListCollections(params: UseListCollectionsParams = {}) {
	const defaultConfig = useConfig();
	const { data: marketplaceConfig } = useMarketplaceConfig();

	const {
		config = defaultConfig,
		marketplaceConfig: paramMarketplaceConfig,
		...rest
	} = params;

	const queryOptions = listCollectionsQueryOptions({
		config,
		marketplaceConfig: paramMarketplaceConfig || marketplaceConfig,
		...rest,
	});

	return useQuery({
		...queryOptions,
	});
}

export { listCollectionsQueryOptions };

export type { FetchListCollectionsParams, ListCollectionsQueryOptions };
