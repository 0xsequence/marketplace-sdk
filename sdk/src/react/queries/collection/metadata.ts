import type { GetContractInfoArgs } from '@0xsequence/metadata';
import { queryOptions } from '@tanstack/react-query';
import type { SdkConfig } from '../../../types';
import {
	getMetadataClient,
	type QueryKeyArgs,
	type ValuesOptional,
} from '../../_internal';

import type { StandardQueryOptions } from '../../types/query';

export interface FetchCollectionParams {
	chainId: number;
	collectionAddress: string;
	config: SdkConfig;
}

/**
 * Fetches collection information from the metadata API
 */
export async function fetchCollection(params: FetchCollectionParams) {
	const { collectionAddress, chainId, config } = params;

	const metadataClient = getMetadataClient(config);

	const result = await metadataClient.getContractInfo({
		chainID: chainId.toString(),
		contractAddress: collectionAddress,
	});

	return result.contractInfo;
}

export type CollectionQueryOptions = ValuesOptional<FetchCollectionParams> & {
	query?: StandardQueryOptions;
};

export function getCollectionQueryKey(params: CollectionQueryOptions) {
	const apiArgs = {
		chainID: String(params.chainId),
		contractAddress: params.collectionAddress,
	} satisfies QueryKeyArgs<GetContractInfoArgs>;

	return ['collection', 'metadata', apiArgs] as const;
}

export function collectionQueryOptions(params: CollectionQueryOptions) {
	const enabled = Boolean(
		params.collectionAddress &&
			params.chainId &&
			params.config &&
			(params.query?.enabled ?? true),
	);

	return queryOptions({
		queryKey: getCollectionQueryKey(params),
		queryFn: () =>
			fetchCollection({
				// biome-ignore lint/style/noNonNullAssertion: The enabled check above ensures these are not undefined
				chainId: params.chainId!,
				// biome-ignore lint/style/noNonNullAssertion: The enabled check above ensures these are not undefined
				collectionAddress: params.collectionAddress!,
				// biome-ignore lint/style/noNonNullAssertion: The enabled check above ensures these are not undefined
				config: params.config!,
			}),
		...params.query,
		enabled,
	});
}
