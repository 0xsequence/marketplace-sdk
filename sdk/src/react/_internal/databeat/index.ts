import type { Auth, Event as DatabeatEvent } from '@databeat/tracker';
import { Databeat } from '@databeat/tracker';

import { useConfig } from '../../hooks';
import type {
	EventType,
	TrackCreateListing,
	TrackCreateOffer,
	TrackSellItems,
	TrackTransactionFailed,
} from './types';

export type EventTypes = keyof typeof EventType;
export type Event = DatabeatEvent<EventTypes>;

export class DatabeatAnalytics extends Databeat<Extract<EventTypes, string>> {
	trackSellItems(args: TrackSellItems) {
		this.track({
			event: 'SELL_ITEMS',
			props: args.props,
			nums: args.nums,
		});
	}

	trackCreateListing(args: TrackCreateListing) {
		this.track({
			event: 'CREATE_LISTING',
			props: args.props,
			nums: args.nums,
		});
	}

	trackCreateOffer(args: TrackCreateOffer) {
		this.track({
			event: 'CREATE_OFFER',
			props: args.props,
			nums: args.nums,
		});
	}

	trackTransactionFailed(args: TrackTransactionFailed) {
		this.track({
			event: 'TRANSACTION_FAILED',
			props: args,
		});
	}
}

export const useAnalytics = () => {
	const config = useConfig();
	const server = 'https://nodes.sequence.app';

	const auth: Auth = {};
	auth.headers = { 'X-Access-Key': config.projectAccessKey };

	return new DatabeatAnalytics(server, auth, {
		defaultEnabled: true,
		initProps: () => {
			return { origin: window.location.origin };
		},
	});
};
