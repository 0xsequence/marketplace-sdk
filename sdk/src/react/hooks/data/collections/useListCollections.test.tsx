import { renderHook, server, waitFor } from '@test';
import { HttpResponse, http } from 'msw';
import type { Address } from 'viem';
import { describe, expect, it } from 'vitest';
import { OrderbookKind } from '../../../_internal';
import {
	createLookupMarketplaceHandler,
	mockConfig,
} from '../../../_internal/api/__mocks__/builder.msw';
import { mockEthCollection } from '../../../_internal/api/__mocks__/metadata.msw';
import { useListCollections } from './useListCollections';

describe('useListCollections', () => {
	it('should fetch collections successfully', async () => {
		const { result } = renderHook(() => useListCollections());

		// Wait for data to be loaded
		await waitFor(() => {
			expect(result.current.data).toBeDefined();
		});

		// Verify the data matches our mock
		expect(result.current.data).toMatchSnapshot();
		expect(result.current.error).toBeNull();
	});

	it('should handle empty collections', async () => {
		// Mock marketplace config with empty collections
		server.use(
			createLookupMarketplaceHandler({
				...mockConfig,
				marketCollections: [],
			}),
		);

		const { result } = renderHook(() => useListCollections());

		await waitFor(() => {
			expect(result.current.data).toBeDefined();
		});

		expect(result.current.data).toMatchSnapshot();
		expect(result.current.error).toBeNull();
	});

	it('should handle disabled queries', async () => {
		let requestMade = false;

		// Mock marketplace config with collection
		server.use(
			createLookupMarketplaceHandler({
				...mockConfig,
				marketCollections: [
					{
						id: 1,
						projectId: 1,
						contractType: 'ERC721',
						chainId: 1,
						itemsAddress:
							'0x1234567890123456789012345678901234567890' as Address,
						feePercentage: 2.5,
						currencyOptions: [],
						bannerUrl: '',
						destinationMarketplace: OrderbookKind.opensea,
						filterSettings: {
							filterOrder: ['Category', 'Level', 'Element'],
							exclusions: [],
						},
						private: false,
						createdAt: new Date('2025-03-16T13:04:16.098Z').toISOString(),
						updatedAt: new Date('2025-03-16T13:04:16.098Z').toISOString(),
					},
				],
			}),
			http.post('*/rpc/Metadata/GetContractInfoBatch', () => {
				requestMade = true;
				return HttpResponse.json({
					contractInfoMap: {
						'0x1234567890123456789012345678901234567890': mockEthCollection,
					},
				});
			}),
		);

		const disabledParams = {
			query: {
				enabled: false,
			},
		};

		const { result } = renderHook(() => useListCollections(disabledParams));

		// For disabled queries, we expect no loading state and no data
		expect(result.current.isLoading).toBe(false);
		expect(result.current.data).toBeUndefined();
		expect(result.current.error).toBeNull();
		expect(requestMade).toBe(false);
	});

	it('should handle multiple collections from different chains', async () => {
		const { result } = renderHook(() => useListCollections());

		await waitFor(() => {
			expect(result.current.data).toBeDefined();
		});

		const set = new Set(result.current.data?.map((c) => c.chainId));
		expect(set.size).toBeGreaterThan(1);
	});
});
