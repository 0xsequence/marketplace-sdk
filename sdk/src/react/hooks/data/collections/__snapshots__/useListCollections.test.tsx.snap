// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useListCollections > should fetch collections successfully 1`] = `
[
  {
    "address": "0x0000000000000000000000000000000000000000",
    "bannerUrl": "https://example.com/market-banner.png",
    "bytecodeHash": "0x1234567890",
    "chainId": 1,
    "contractType": "ERC721",
    "createdAt": "2025-03-16T13:04:16.098Z",
    "currencyOptions": [
      "0x0000000000000000000000000000000000000000",
      "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
    ],
    "deployed": true,
    "destinationMarketplace": "sequence_marketplace_v2",
    "extensions": {
      "blacklist": false,
      "categories": [
        "Mock",
      ],
      "description": "A mock collection for testing",
      "featureIndex": 0,
      "featured": true,
      "link": "https://example.com",
      "ogImage": "https://example.com/image.png",
      "ogName": "Mock Collection",
      "originAddress": "0x0000000000000000000000000000000000000000",
      "originChainId": 1,
      "verified": true,
      "verifiedBy": "0x",
    },
    "feePercentage": 3.5,
    "filterSettings": {
      "exclusions": [
        {
          "condition": "SPECIFIC_VALUE",
          "key": "Type",
          "value": "Sample",
        },
      ],
      "filterOrder": [
        "Type",
        "Rarity",
      ],
    },
    "id": 1,
    "itemsAddress": "0x0000000000000000000000000000000000000000",
    "logoURI": "https://example.com/logo.png",
    "marketplaceType": "market",
    "name": "Mock Collection",
    "private": false,
    "projectId": 1,
    "source": "https://example.com/source",
    "status": "AVAILABLE",
    "symbol": "MOCK",
    "type": "ERC721",
    "updatedAt": "2025-05-01T04:39:56.936Z",
  },
  {
    "address": "0x1234567890123456789012345678901234567890",
    "bannerUrl": "https://example.com/collection-banner.png",
    "bytecodeHash": "0x1234567890",
    "chainId": 137,
    "contractType": "ERC1155",
    "createdAt": "2025-03-16T13:04:16.098Z",
    "currencyOptions": [
      "0x0000000000000000000000000000000000000000",
    ],
    "deployed": true,
    "destinationMarketplace": "opensea",
    "extensions": {
      "blacklist": false,
      "categories": [
        "Mock",
      ],
      "description": "A mock collection for testing",
      "featureIndex": 0,
      "featured": true,
      "link": "https://example.com",
      "ogImage": "https://example.com/image.png",
      "ogName": "Mock Collection",
      "originAddress": "0x0000000000000000000000000000000000000000",
      "originChainId": 1,
      "verified": true,
      "verifiedBy": "0x",
    },
    "feePercentage": 2.5,
    "filterSettings": {
      "exclusions": [
        {
          "condition": "ENTIRE_KEY",
          "key": "Category",
        },
        {
          "condition": "SPECIFIC_VALUE",
          "key": "Level",
          "value": "Legendary",
        },
      ],
      "filterOrder": [
        "Category",
        "Level",
        "Element",
      ],
    },
    "id": 2,
    "itemsAddress": "0x1234567890123456789012345678901234567890",
    "logoURI": "https://example.com/logo.png",
    "marketplaceType": "market",
    "name": "Mock Collection",
    "private": false,
    "projectId": 1,
    "source": "https://example.com/source",
    "status": "AVAILABLE",
    "symbol": "MOCK",
    "type": "ERC721",
    "updatedAt": "2025-05-01T04:39:56.936Z",
  },
  {
    "address": "0x0000000000000000000000000000000000000000",
    "bannerUrl": "https://example.com/shop-banner.png",
    "bytecodeHash": "0x1234567890",
    "chainId": 1,
    "createdAt": "2025-03-16T13:04:16.098Z",
    "customTokenIds": [],
    "deployed": true,
    "extensions": {
      "blacklist": false,
      "categories": [
        "Mock",
      ],
      "description": "A mock collection for testing",
      "featureIndex": 0,
      "featured": true,
      "link": "https://example.com",
      "ogImage": "https://example.com/image.png",
      "ogName": "Mock Collection",
      "originAddress": "0x0000000000000000000000000000000000000000",
      "originChainId": 1,
      "verified": true,
      "verifiedBy": "0x",
    },
    "id": 1,
    "itemsAddress": "0x0000000000000000000000000000000000000000",
    "logoURI": "https://example.com/logo.png",
    "marketplaceType": "shop",
    "name": "Mock Collection",
    "private": false,
    "projectId": 1,
    "saleAddress": "0x0000000000000000000000000000000000000000",
    "saleBannerUrl": "https://example.com/shop-sale-banner.png",
    "source": "https://example.com/source",
    "status": "AVAILABLE",
    "symbol": "MOCK",
    "tokenIds": [
      "1",
      "2",
      "3",
    ],
    "type": "ERC721",
    "updatedAt": "2025-05-01T04:39:56.936Z",
  },
  {
    "address": "0x1234567890123456789012345678901234567890",
    "bannerUrl": "https://example.com/polygon-shop-banner.png",
    "bytecodeHash": "0x1234567890",
    "chainId": 137,
    "createdAt": "2025-03-16T13:04:16.098Z",
    "customTokenIds": [],
    "deployed": true,
    "extensions": {
      "blacklist": false,
      "categories": [
        "Mock",
      ],
      "description": "A mock collection for testing",
      "featureIndex": 0,
      "featured": true,
      "link": "https://example.com",
      "ogImage": "https://example.com/image.png",
      "ogName": "Mock Collection",
      "originAddress": "0x0000000000000000000000000000000000000000",
      "originChainId": 1,
      "verified": true,
      "verifiedBy": "0x",
    },
    "id": 2,
    "itemsAddress": "0x1234567890123456789012345678901234567890",
    "logoURI": "https://example.com/logo.png",
    "marketplaceType": "shop",
    "name": "Mock Collection",
    "private": false,
    "projectId": 1,
    "saleAddress": "0x1234567890123456789012345678901234567890",
    "saleBannerUrl": "https://example.com/polygon-shop-sale-banner.png",
    "source": "https://example.com/source",
    "status": "AVAILABLE",
    "symbol": "MOCK",
    "tokenIds": [
      "4",
      "5",
      "6",
    ],
    "type": "ERC721",
    "updatedAt": "2025-05-01T04:39:56.936Z",
  },
]
`;

exports[`useListCollections > should handle empty collections 1`] = `
[
  {
    "address": "0x0000000000000000000000000000000000000000",
    "bannerUrl": "https://example.com/shop-banner.png",
    "bytecodeHash": "0x1234567890",
    "chainId": 1,
    "createdAt": "2025-03-16T13:04:16.098Z",
    "customTokenIds": [],
    "deployed": true,
    "extensions": {
      "blacklist": false,
      "categories": [
        "Mock",
      ],
      "description": "A mock collection for testing",
      "featureIndex": 0,
      "featured": true,
      "link": "https://example.com",
      "ogImage": "https://example.com/image.png",
      "ogName": "Mock Collection",
      "originAddress": "0x0000000000000000000000000000000000000000",
      "originChainId": 1,
      "verified": true,
      "verifiedBy": "0x",
    },
    "id": 1,
    "itemsAddress": "0x0000000000000000000000000000000000000000",
    "logoURI": "https://example.com/logo.png",
    "marketplaceType": "shop",
    "name": "Mock Collection",
    "private": false,
    "projectId": 1,
    "saleAddress": "0x0000000000000000000000000000000000000000",
    "saleBannerUrl": "https://example.com/shop-sale-banner.png",
    "source": "https://example.com/source",
    "status": "AVAILABLE",
    "symbol": "MOCK",
    "tokenIds": [
      "1",
      "2",
      "3",
    ],
    "type": "ERC721",
    "updatedAt": "2025-05-01T04:39:56.936Z",
  },
  {
    "address": "0x1234567890123456789012345678901234567890",
    "bannerUrl": "https://example.com/polygon-shop-banner.png",
    "bytecodeHash": "0x1234567890",
    "chainId": 137,
    "createdAt": "2025-03-16T13:04:16.098Z",
    "customTokenIds": [],
    "deployed": true,
    "extensions": {
      "blacklist": false,
      "categories": [
        "Mock",
      ],
      "description": "A mock collection for testing",
      "featureIndex": 0,
      "featured": true,
      "link": "https://example.com",
      "ogImage": "https://example.com/image.png",
      "ogName": "Mock Collection",
      "originAddress": "0x0000000000000000000000000000000000000000",
      "originChainId": 1,
      "verified": true,
      "verifiedBy": "0x",
    },
    "id": 2,
    "itemsAddress": "0x1234567890123456789012345678901234567890",
    "logoURI": "https://example.com/logo.png",
    "marketplaceType": "shop",
    "name": "Mock Collection",
    "private": false,
    "projectId": 1,
    "saleAddress": "0x1234567890123456789012345678901234567890",
    "saleBannerUrl": "https://example.com/polygon-shop-sale-banner.png",
    "source": "https://example.com/source",
    "status": "AVAILABLE",
    "symbol": "MOCK",
    "tokenIds": [
      "4",
      "5",
      "6",
    ],
    "type": "ERC721",
    "updatedAt": "2025-05-01T04:39:56.936Z",
  },
]
`;
